WordPress/src/main/java/org/wordpress/android/ui/ShareIntentReceiverActivity.java

```diff
@@ -4,233 +4,158 @@
-import org.wordpress.android.ui.posts.EditPostActivity;
-import org.wordpress.android.ui.prefs.AppPrefs;
 import org.wordpress.android.util.AnalyticsUtils;
 import org.wordpress.android.util.FluxCUtils;
 import org.wordpress.android.util.PermissionUtils;
-import org.wordpress.android.util.SiteUtils;
-import org.wordpress.android.util.ToastUtils;
 import org.wordpress.android.util.WPPermissionUtils;
 
 import java.util.ArrayList;
 import java.util.HashMap;
-import java.util.List;
 import java.util.Map;
-
 import javax.inject.Inject;
 
 /**
  * An activity to handle share intents, since there are multiple actions possible.
- * If there are multiple blogs, it lets the user choose which blog to share to.
- * It lists what actions that the user can perform and redirects them to the activity,
- * along with the content passed in the intent
+ * If the user is not logged in, redirects the user to the LoginFlow. When the user is logged in,
+ * displays ShareIntentReceiverFragment. The fragment lets the user choose which blog to share to.
+ * Moreover it lists what actions the user can perform and redirects the user to the activity,
+ * along with the content passed in the intent.
  */
-public class ShareIntentReceiverActivity extends AppCompatActivity implements OnItemSelectedListener {
-    public static final String SHARE_LAST_USED_BLOG_ID_KEY = "wp-settings-share-last-used-text-blogid";
-    public static final String SHARE_LAST_USED_ADDTO_KEY = "wp-settings-share-last-used-image-addto";
+public class ShareIntentReceiverActivity extends AppCompatActivity implements ShareIntentFragmentListener {
 
-    private static final int ADD_TO_NEW_POST = 0;
-    private static final int ADD_TO_MEDIA_LIBRARY = 1;
+    public static final String SHARE_LAST_USED_BLOG_ID_KEY = "wp-settings-share-last-used-text-blogid";
+    private static final String KEY_SELECTED_SITE_LOCAL_ID = "KEY_SELECTED_SITE_LOCAL_ID";
+    private static final String KEY_SHARE_ACTION_ID = "KEY_SHARE_ACTION_ID";
 
     @Inject AccountStore mAccountStore;
     @Inject SiteStore mSiteStore;
 
-    private Spinner mBlogSpinner;
-    private RadioGroup mActionGroup;
-    private String mSiteNames[];
-    private int mSiteIds[];
-    private int mSelectedSiteLocalId;
-    private int mActionIndex;
+    private int mClickedSiteLocalId;
+    private String mShareActionName;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         ((WordPress) getApplication()).component().inject(this);
+        setContentView(R.layout.share_intent_receiver_activity);
 
-        setContentView(R.layout.share_intent_receiver_dialog);
-
-        TextView blogSpinnerTitle = (TextView) findViewById(R.id.blog_spinner_title);
-        mBlogSpinner = (Spinner) findViewById(R.id.blog_spinner);
-        mActionGroup = (RadioGroup) findViewById(R.id.share_actions);
-
-        initSiteLists();
-
-        if (mSiteNames == null) {
```

Comment:
This (ugly) test, was checking if the user had at least one visible site, and if not, we were closing the share activity with a toast message:

```
ToastUtils.showToast(getBaseContext(), R.string.cant_share_no_visible_blog, ToastUtils.Duration.LONG);
```

Also, some wordpress.com users don't have any site, in this situation, the picker crash after tapping the "add to new post" button:

```
 java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
     at java.util.ArrayList.get(ArrayList.java:411)
     at org.wordpress.android.ui.main.SitePickerAdapter.getItem(SitePickerAdapter.java:188)
     at org.wordpress.android.ui.main.SitePickerAdapter.getSelectedItemLocalId(SitePickerAdapter.java:329)
     at org.wordpress.android.ui.ShareIntentReceiverFragment$1.onClick(ShareIntentReceiverFragment.java:128)
     at android.view.View.performClick(View.java:5609)
     at android.view.View$PerformClick.run(View.java:22259)
     at android.os.Handler.handleCallback(Handler.java:751)
     at android.os.Handler.dispatchMessage(Handler.java:95)
     at android.os.Looper.loop(Looper.java:154)
     at android.app.ActivityThread.main(ActivityThread.java:6077)
     at java.lang.reflect.Method.invoke(Native Method)
     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
```

Reply:
The check is in the ShareIntentReceiverFragment -> createSiteAdapter() -> mAdapter.getItemCount() == 0. But you are right, there is a bug. The mAdapter.getItemCount() returns 1 if there is a header and no sites.